libzfs_core/_libzfs_core.py

```diff
@@ -0,0 +1,683 @@
+# Copyright 2015 ClusterHQ. See LICENSE file for details.
+
+"""
+Python wrappers for libzfs_core interfaces.
+
+As a rule, there is a Python function for each C function.
+The signatures of the Python functions generally follow those of the
+functions, but the argument types are natural to Python.
+nvlists are wrapped as dictionaries or lists depending on their usage.
+Some parameters have default values depending on typical use for
+increased convenience.  Output parameters are not used and return values
+are directly returned.  Error conditions are signalled by exceptions
+rather than by integer error codes.
+"""
+
+import errno
+import threading
+from . import exceptions
+from . import _error_translation as xlate
+from .bindings import libzfs_core
+from ._constants import MAXNAMELEN
+from ._nvlist import nvlist_in, nvlist_out
+
+
+def lzc_create(name, is_zvol=False, props=None):
+    '''
+    Create a ZFS filesystem or a ZFS volume ("zvol").
+
+    :param bytes name: a name of the dataset to be created.
+    :param bool is_zvol: whether to create a zvol (false by default).
+    :param props: a `dict` of ZFS dataset property name-value pairs (empty by default).
+    :type props: dict of bytes:Any
+
+    :raises FilesystemExists: if a dataset with the given name already exists.
+    :raises ParentNotFound: if a parent dataset of the requested dataset does not exist.
+    :raises PropertyInvalid: if one or more of the specified properties is invalid
+                             or has an invalid type or value.
+
+    Wraps ``int lzc_create(const char *fsname, dmu_objset_type_t type, nvlist_t *props)``.
+    '''
+    if props is None:
+        props = {}
+    if is_zvol:
+        ds_type = _lib.DMU_OST_ZVOL
+    else:
+        ds_type = _lib.DMU_OST_ZFS
```

Comment:
Those have been the only types so far and I haven't heard of any plans for a new type.
However, I do not object against making the interface future-proof.

One slight problem with the dynamic lookup approach is that the full list of types at the moment looks like this:
```
typedef enum dmu_objset_type {
        DMU_OST_NONE,
        DMU_OST_META,
        DMU_OST_ZFS,
        DMU_OST_ZVOL,
        DMU_OST_OTHER,                  /* For testing only! */
        DMU_OST_ANY,                    /* Be careful! */
        DMU_OST_NUMTYPES
} dmu_objset_type_t;
```
I am sure that the kernel code would be able to gracefully handle the wrong types (or maybe not :smile:).
But that would be another `EINVAL` to reverse-map and that would be quite painful...
I guess that if I would do a post-error validation of the type against the hardcoded list of supported types, then it would be as a good as hardcoded mapping of the input to the C type (without the dynamic lookup trick).